File: janMarkusMilan/assignments/treesDFS/lc112/lc112.js
// Problem Statement #

// Given a binary tree and a number ‘S’, find if the tree has a path from root-to-leaf such that the sum of all the node values of that path equals ‘S’.

/*
Edge Case:
  1. If !node -> return false

Root to leaf: Use DFS

General idea of recursive implementation:
  function recursive (node, count) {
    if (!node) return 0;
    count++;
    return recursive(node, count);
  }

What to take note of:
  1. Since the question is checking for a sum, then we need to pass that as an argument in our recursive function (including the node)
  2. For each node we reach, we add the value contained inside to our sum
  3. If our sum equals the target sum at any point, then just return true
  4. If we reach the end of a branch, we check if our sum equals the target sum
*/

class TreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

const has_path = function (root, sum) {
  // TODO: Write your code here
  function DFS(node, currentSum) {
    if (!node) {
      return false;
    }
    currentSum += node.value;
    if (!node.left && !node.right) {
      return currentSum === sum;
    }
    if (DFS(node.left, currentSum) || DFS(node.right, currentSum)) {
      return true;
    }
    return false;
  }

  return DFS(root, 0);
};

/*
        (12)
        /  \
      (7)  (1)
      /    / \
    (9)  (10)(5)
*/

var root = new TreeNode(12);
root.left = new TreeNode(7);
root.right = new TreeNode(1);
root.left.left = new TreeNode(9);
root.right.left = new TreeNode(10);
root.right.right = new TreeNode(5);
console.log(`Tree has path: ${has_path(root, 23)}`);
console.log(`Tree has path: ${has_path(root, 16)}`);
console.log(`Tree has path: ${has_path(root, 12)}`);
console.log(`Tree has path: ${has_path(root, 0)}`);
console.log(`Tree has path: ${has_path(root, null)}`);
console.log(`Tree has path: ${has_path(root, 28)}`);

File: roseWong/assignments/treesDFS/lc112/lc112.py
# Problem Statement #


# Given a binary tree and a number ‘S’, find if the tree has a path from root-to-leaf such that the
# sum of all the node values of that path equals ‘S’.

# Start at the root and traverse down each branch until I hit a leave.
# At every level, subract the value of the node from the sum (S - val of the node) that I am on.
# At the end of each branch (when a leaf is encountered), check to see if the value of sum equals
# the value of the leaf node.  If equal, return true, else return false
# This code uses a recursive function.

# If each function call of recursive algorithm takes O(m) space, and m is made up of the parameters of function,
# local variables in the function, and return address of caller function. 
# and if the maximum depth of recursion tree is 'n'
# then space complexity of recursive algorithm would be O(nm)
# time complexity is O(2^n), where n is height of our tree.

# Key learnings:  first do the edge case, then do the base case, then complete with the recursive logic.

class TreeNode:
  def __init__(self, val, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


def has_path(root, sum):

  if not root :
    return False
  
  if not root.left and not root.right:  #terminating case
    return sum == root.val

  return has_path(root.left, sum-root.val) or has_path(root.right, sum-root.val)


def main():

  root = TreeNode(12)
  root.left = TreeNode(7)
  root.right = TreeNode(1)
  root.left.left = TreeNode(9)
  root.right.left = TreeNode(10)
  root.right.right = TreeNode(5)
  print("Tree has path: " + str(has_path(root, 23)))
  print("Tree has path: " + str(has_path(root, 16)))


main()
